### PR TITLE
Require C++17 standard

### PR DESCRIPTION
**Pull request by aprokop**
I disabled several Clang-Tidy checks:

- modernize-use-no-discard
- modernize-use-concat-namespaces
- modernize-unary-static-assert

### CODE REVIEW

File: .jenkins/nightly.groovy [hunk 1]
```diff
@@ -17,7 +17,7 @@ pipeline {
                     }
                     environment {
                         CTEST_OPTIONS = '--timeout 180 --no-compress-output -T Test'
-                        CMAKE_OPTIONS = '-D CMAKE_BUILD_TYPE=Release -D CMAKE_CXX_STANDARD=17 -D CMAKE_CXX_EXTENSIONS=OFF'
```
Comment (dalg24) [inline .jenkins/nightly.groovy:20]: Why did you remove?  Relying on defaults rather being explicit about it?

File: test/tstCompileOnlyWeightedEdges.cpp [hunk 2]
```diff
@@ -14,57 +14,54 @@
...
-#define STATIC_ASSERT(cond) static_assert(cond, "")
-
 KOKKOS_FUNCTION constexpr bool test_directed_edges()
 {
   using ArborX::Details::DirectedEdge;
 
-  STATIC_ASSERT((DirectedEdge{0, 1, 2.f}.weight == 2.f));
-  STATIC_ASSERT((DirectedEdge{0, 1, 2.f}.source() == 0));
-  STATIC_ASSERT((DirectedEdge{0, 1, 2.f}.target() == 1));
+  static_assert((DirectedEdge{0, 1, 2.f}.weight == 2.f));
```
Comment (dalg24) [inline test/tstCompileOnlyWeightedEdges.cpp:21]: Drop the extra parentheses

**Review by dalg24 — COMMENTED** [review]
I'd recommend doing the cleanups elsewhere

**Comment by aprokop** [comment]
> I'd recommend doing the cleanups elsewhere

I'm OK with that. I would do `static_assert` in a separate PR. I still want to keep `static constexpr int` for the sanity check.

File: .jenkins/nightly.groovy (hunk 1 above)
Comment (aprokop) [reply .jenkins/nightly.groovy:20]: My assumption was that it was not so much relying on the exact standard, rather it was requiring at least C++17. As this PR now provides C++17, it seemed no longer necessary.

Comment (dalg24) [reply .jenkins/nightly.groovy]: This variable was enforcing consistent options on Kokkos and ArborX.  I actually think it is a good idea.  Assume we of a C++20 build for instance.

Comment (aprokop) [reply .jenkins/nightly.groovy]: I restored it back.

File: docker/README.md [hunk 3]
```diff
@@ -14,5 +14,5 @@ services:
       args:
         - BASE=nvidia/cuda:11.4.2-devel-ubuntu20.04
         - KOKKOS_VERSION=3.6.00
-        - KOKKOS_OPTIONS=-DCMAKE_CXX_STANDARD=17 -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=ON -DKokkos_ARCH_SNB=ON -DKokkos_ARCH_VOLTA70=ON
+        - KOKKOS_OPTIONS=-DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=ON -DKokkos_ARCH_SNB=ON -DKokkos_ARCH_VOLTA70=ON
```
Comment (masterleinad) [inline docker/README.md:17]: I would leave this since it enables some more `Kokkos` features.

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
CUDA+nvcc testing fails due using GCC 7 (see kokkos/kokkos#5157). Need to update the gcc compiler in the images.

**Comment by aprokop** [comment]
Tons of warnings from Clang-Tidy for `static_assert(blah, "");` [modernize-unary-static-assert]. Disabled in this PR, will reenable in the `static_assert` cleanup PR.

**Review by dalg24 — APPROVED** [review]
